---
title: "GxEScanR: An R Package to Detect GxE Interactions in a Genome-wide Association Study"
bibliography: vignettes/library.bib
csl: vignettes/nature.csl
output: 
  github_document:
    pandoc_args: --webtex
always_allow_html: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = T,
  collapse = TRUE,
  comment = "#>"
)

knitr::opts_knit$set(root.dir = './data')
```
*Version: 1.0*  
*October 2018*   

##### Developed by:   

  - John Morrison, M.S. (JMorr@usc.edu)  
  - Andre E. Kim, Ph.D. (Andreeki@usc.edu)  
  - Jim Gauderman, Ph.D. (JimG@usc.edu)  

##### Acknowledgements
Supported in part by NCI grants PO1-CA196569 and RO1-CA201407 and NIEHS grant P30-ES007048.  

##### Reference, software
1. Morrison, J, Kim AE, Gauderman J.  GxEScanR:  An R package to detect GxE interactions in a genomewide association study,  Version 1.0.  University of Southern California, Los Angeles; October, 2018.  http://Github.com/USCBiostats/GxEScanR

##### References, methods
2. Gauderman J, Mukherjee B, Aschard H, Hsu L, Lewinger JP, Patel C, Witte J, Amos C, Tai C, Conti D, Torgerson D, Lee S, Chatterjee N. Update on the State of the Science for Analytical Methods for Gene-Environment Interactions.  Am J Epidemiol 186:762-70, 2017.  

3. Gauderman J, Zhang P, Morrison J, Lewinger JP.  Finding Novel Genes by Testing 
GxE Interactions in a Genomewide Association Study.  Genetic Epidemiology,  37:603-613, 2013.  
<br>


*Thank you for using GxEScanR for your GxE interation analysis.*   
<br>
*We would appreciate hearing about any errors you might find or comments you might have to improve the program.  Please e-mail questions/comments to Jim Gauderman.*  

------

# General Description

GxEScanR is an R package that performs a genomewide scan for gene-environment (GxE) interaction for a disease trait.  The program implements the traditional test of GxE interaction in a case-control sample, the case-only test of GxE interaction, the 2-degree-of-freedom (df) joint test of G and GxE, and efficient 2-step methods for tests of either GxE or joint G & GxE[@Gauderman2017].  As a byproduct of the available tests, GxEScanR also produces results for the standard marginal G scan, i.e. the standard test of each SNP conducted in a GWAS.

The ‘environment’ factor E in the GxE scans may be either binary or continuous and can be an exogenous exposure variable (e.g., sunlight, air pollution), personal exposure (e.g., smoking, dietary fat), or other personal characteristic (e.g., sex, age, candidate gene).  GxEScanR will test GxE interaction with measured and/or imputed SNPs on autosomal chromosomes and will control the family-wise error rate (FWER) at a user-defined level (e.g. 5%).

GxEScanR utilizes C++ (via RCpp) to improve computational speed.  The program also reads binary-formatted imputed SNP files which further improves speed.  A companion program, [BinaryDosage](http://Github.com/USCBiostats/BinaryDosage), is available to convert VCF or IMPUTE2 imputed SNP files into the binary format required by GxEScanR.  The BinaryDosage program should be used before using GxEScanR.


### Functions
- **GetBinaryDosageInfo** - Create R List containing BDose file attributes (required for *GxEScan*)
- **GxEScan** - Run GWAS/GWIS (See below)

### Program Outputs
GxEScanR produces a single text-formatted output file that includes the SNP identifier and the effect estimate and test statistic for all of the tests described below. 

##### Output statistics
-	**BetaG/zG**: Marginal G (CC_DG) association  
-	**BetaGxE/zGxE**: GxE (CC_GxE) interaction  
- **Chi2df**: Joint G, GxE (CC_2df) association  
-	**Beta_Case/z_Case**: Case-only G-E association (Case_GE)  
-	**Beta_Control/z_Control**: Control-only G-E association (Cntl_GE)  
-	**Beta_GE/z_GE**: Overall G-E association in cases and controls combined (CC_GE)  

These statistics can be used to generate QQplots and/or Manhattan plots using your favorite R program or other software.  A companion R package that will generate QQplots and Manhattan plots for all of the tests generated by GxEScanR is currently under construction.  

With additional R coding (at this point on your own) the above statistics can be used to generate the CC_3df test (see Section 2.5) or any of the 2-step procedures (see Section 2.6).  A companion R Package that will read GxEScanR output and generate the 3-df and 2-step tests described above is currently under construction.  

### Installation
1. Install the [devtools](https://github.com/hadley/devtools) package
2. Install the [GxEScanR](https://github.com/USCbiostats/GxEScanR) package directly from the USCbiostats repository on GitHub: 

``` {r, eval = F}
library(devtools)
install_github("USCbiostats/GxEScanR")

library(GxEScanR)
```

### Example

Use example BDose and dummy covariate file provided in GitHub repository.

##### Run GxEScan
