---
title: "GxEScanR: Performs GWAS/GWEIS scans using BinaryDosage files"
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# GxEScanR

GxEScanR is designed to efficiently run genome-wide association study (GWAS) and genome-wide by environmental interaction study (GWEIS) scans using imputed genotypes
stored in the BinaryDosage format. The phenotype to be analyzed can either be a
continuous or binary trait. The GWEIS scan performs multiple tests that can be
used in two-step methods.

### Development version

The development version has an improved convergence routine. It is not
required but it is recommended to use the latest development version of the
BinaryDosage package. Click 
[here](https://github.com/USCbiostats/BinaryDosage/tree/htslib) for instructions

## Installation

To install the development version from [GitHub](https://github.com/) run
the following code.

``` r
# Uncomment the following line if devtools is not installed.
# install.packages("devtools")
devtools::install_github("USCbiostats/GxEScanR@develop")
```
## Example

Needs to be written.
