# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' Function to produce summary of a binary dosage file used by GxEScan
#' 
#' Function to produce summary of a binary dosage file used by GxEScan
#' 
#' @param binaryDosageFilesname
#' Name of file with genetic data, normally ends with .bdosage
#' @param snSub
#' Number of subjects in the genetic data file. This is needed for formats 1, 2 and 3.
#' This value is returned for format 4 in the results
#' @param snSNPs
#' Number of SNPs in the genetic data file. This is needed for formats 1, 2 and 3.
#' This value is returned for format 4 in the results
#' @return
#' List with data needed by GxEScan
#' @export
GetBinaryDosageInformation <- function(binaryDosageFilename, snSub, snSNPs) {
    .Call('_GxEScanR_GetBinaryDosageInformation', PACKAGE = 'GxEScanR', binaryDosageFilename, snSub, snSNPs)
}

#' Function to fit models scanning over genotypes
#' 
#' Function to fit selected models over genotypes. Results from
#' these models are then processed by GxETest to display the results
#' for these one step tests along with two step tests derived from
#' the one step tests.
#' 
#' @param subjectData
#' List returned from SubsetSubjects
#' @param geneticInfo
#' List returned from one of the functions to get the required information
#' about the source of genetic data.
#' @param outputFilename
#' Name of output file
#' @param skippedFilename
#' Name of file to write info about SNPs that were skipped. If this is blank
#' no file is written. If this is the same as outputFilename, the skipped SNPs
#' are written to the output file along with NA for all tests.
#' @param minMaf
#' Minimum minor allele frequency. Must be between 0.0001 and 0.25
#' @param geCutoff
#' p-value cut off value for logistic G|E test to fit polytomous models.
#' @return
#' 0 success
#' 1 failure
#' @export
GxEScanC <- function(subjectData, geneticInfo, outputFilename, skippedFilename, minMaf, geCutoff) {
    .Call('_GxEScanR_GxEScanC', PACKAGE = 'GxEScanR', subjectData, geneticInfo, outputFilename, skippedFilename, minMaf, geCutoff)
}

#' Function to fit models scanning over a subset of genotypes
#' 
#' Function to fit selected models over a subset of genotypes. Results from
#' these models are then processed by GxETest to display the results
#' for these one step tests along with two step tests derived from
#' the one step tests.
#' 
#' @param subjectData
#' List returned from SubsetSubjects
#' @param geneticInfo
#' List returned from one of the functions to get the required information
#' about the source of genetic data.
#' @param outputFilename
#' Name of output file
#' @param skippedFilename
#' Name of file to write info about SNPs that were skipped. If this is blank
#' no file is written. If this is the same as outputFilename, the skipped SNPs
#' are written to the output file along with NA for all tests.
#' @param minMaf
#' Minimum minor allele frequency. Must be between 0.0001 and 0.25
#' @param geCutoff
#' p-value cut off value for logistic G|E test to fit polytomous models.
#' @param snpIndices
#' Indices of SNP locations in geneticInfo to be used in analysis
#' @return
#' 0 success
#' 1 failure
#' @export
GxEScanCSubset <- function(subjectData, geneticInfo, outputFilename, skippedFilename, minMaf, geCutoff, snpIndices) {
    .Call('_GxEScanR_GxEScanCSubset', PACKAGE = 'GxEScanR', subjectData, geneticInfo, outputFilename, skippedFilename, minMaf, geCutoff, snpIndices)
}

#' Function to display the results from the scans performed by GxEScan
#' 
#' Function to display the results from the scans performed by GxEScan.
#' The results include Manhattan and qq-plot along with tables of the
#' top hits for each test. The results are saved to an Excel spreadsheet.
#' 
#' @return
#' 0 success
#' 1 failure
#' @export
GxETest <- function() {
    .Call('_GxEScanR_GxETest', PACKAGE = 'GxEScanR')
}

#' Function to calculate allele frequencies
#' 
#' Function to calculate allele frequencies
#' 
#' @param subjectData
#' List returned from SubsetSubjects
#' @param geneticInfo
#' List returned from one of the functions to get the required information
#' about the source of genetic data.
#' @param outputFilename
#' Name of output file
#' @return
#' 0 success
#' 1 failure
#' @export
GxEScanFreqC <- function(subjectData, geneticInfo, outputFilename) {
    .Call('_GxEScanR_GxEScanFreqC', PACKAGE = 'GxEScanR', subjectData, geneticInfo, outputFilename)
}

#' Function to calculate allele frequencies
#' 
#' Function to calculate allele frequencies
#' 
#' @param subjectData
#' List returned from SubsetSubjects
#' @param geneticInfo
#' List returned from one of the functions to get the required information
#' about the source of genetic data.
#' @param outputFilename
#' Name of output file
#' @return
#' 0 success
#' 1 failure
#' @export
GxEScanFreqC2 <- function(subjectData, geneticInfo, outputFilename) {
    .Call('_GxEScanR_GxEScanFreqC2', PACKAGE = 'GxEScanR', subjectData, geneticInfo, outputFilename)
}

#' Function to convert and Impute 2 style file to a binary dosage file
#' 
#' Function to convert and Impute 2 style file to a binary dosage file
#' 
#' @param imp2Info
#' List containing information of Impute 2 file, returned from GetI2Info
#' @param filename
#' Name of binary dosage file to create
#' @param format
#' Primary format of binary dosage file
#' @param subformat
#' Secondary format of binary dosage file
#' @return
#' 0 success
#' 1 failure
#' @export
Imp2toBDC <- function(imp2Info, filename, format, subformat) {
    .Call('_GxEScanR_Imp2toBDC', PACKAGE = 'GxEScanR', imp2Info, filename, format, subformat)
}

#' Function to produce summary of a plink binary genetic file used by GxEScan
#' 
#' Function to produce summary of a plink binary genetic file used by GxEScan
#' 
#' @param geneticFile
#' Name of file with genetic data, normally ends with .bed
#' @param mapFile
#' Name of map file associated with genetic data file, normally ends with .bim
#' @param familyFile
#' Name of family file associated with genetic data file, normally ends with .fam
#' @return
#' List with data needed by GxEScan
#' @export
PlinkBinaryInfo <- function(geneticFile, mapFile, familyFile) {
    .Call('_GxEScanR_PlinkBinaryInfo', PACKAGE = 'GxEScanR', geneticFile, mapFile, familyFile)
}

OpenGxEOutFile <- function(filename) {
    .Call('_GxEScanR_OpenGxEOutFile', PACKAGE = 'GxEScanR', filename)
}

AppendGxEResults <- function(filename, snpID, chromosome, location, refAllele, altAllele, numSub, numCases, logLike, estimates, length, sigmaE) {
    .Call('_GxEScanR_AppendGxEResults', PACKAGE = 'GxEScanR', filename, snpID, chromosome, location, refAllele, altAllele, numSub, numCases, logLike, estimates, length, sigmaE)
}

InitializeLRMod <- function(numRow, numCol, y, xl, beta, score, w, wInv, yp, zt, k, ql, rtl, abx, expabx, expabxp1, expitabx, logLikelihood) {
    .Call('_GxEScanR_InitializeLRMod', PACKAGE = 'GxEScanR', numRow, numCol, y, xl, beta, score, w, wInv, yp, zt, k, ql, rtl, abx, expabx, expabxp1, expitabx, logLikelihood)
}

FitLRMod <- function(n, p, y, xl, xr, beta0, score0, w, wInv, yp0, zt0, k0, ql, rtl, abx, expabx, expabxp1, expitabx, yp, zt, k, bt, xrw, beta, score, zb, bb, h, rtr, t, qr, rbr, logLikelihood) {
    .Call('_GxEScanR_FitLRMod', PACKAGE = 'GxEScanR', n, p, y, xl, xr, beta0, score0, w, wInv, yp0, zt0, k0, ql, rtl, abx, expabx, expabxp1, expitabx, yp, zt, k, bt, xrw, beta, score, zb, bb, h, rtr, t, qr, rbr, logLikelihood)
}

ScanGenes <- function(n, p, y, xl, xr, numSNPs, minMAF, beta0, score0, w, wInv, yp0, zt0, k0, ql, rtl, abx, expabx, expabxp1, expitabx, yp, zt, k, bt, xrw1, beta1, score1, zb1, bb1, h1, rtr1, t1, qr1, rbr1, logLikelihood1, xrw2, beta2, score2, zb2, bb2, h2, rtr2, t2, qr2, rbr2, logLikelihood2, xr1, xr2, logLikelihood0, logLikelihoods, estimates) {
    .Call('_GxEScanR_ScanGenes', PACKAGE = 'GxEScanR', n, p, y, xl, xr, numSNPs, minMAF, beta0, score0, w, wInv, yp0, zt0, k0, ql, rtl, abx, expabx, expabxp1, expitabx, yp, zt, k, bt, xrw1, beta1, score1, zb1, bb1, h1, rtr1, t1, qr1, rbr1, logLikelihood1, xrw2, beta2, score2, zb2, bb2, h2, rtr2, t2, qr2, rbr2, logLikelihood2, xr1, xr2, logLikelihood0, logLikelihoods, estimates)
}

GetLocations <- function(x, y, fileSize, bufferSize) {
    .Call('_GxEScanR_GetLocations', PACKAGE = 'GxEScanR', x, y, fileSize, bufferSize)
}

WriteLocations <- function(y, x) {
    .Call('_GxEScanR_WriteLocations', PACKAGE = 'GxEScanR', y, x)
}

GetSections <- function(y, snpSection, fileLocation, snpLocation, bufferSize) {
    .Call('_GxEScanR_GetSections', PACKAGE = 'GxEScanR', y, snpSection, fileLocation, snpLocation, bufferSize)
}

ReadSNP <- function(snpNumber, subjectNumber, filename, format, numSub, numSNPs, bufferSize, buffer, sections, snpSection, fileLocation, snpLocation, currentSection, dosage, p0, p1, p2, values) {
    .Call('_GxEScanR_ReadSNP', PACKAGE = 'GxEScanR', snpNumber, subjectNumber, filename, format, numSub, numSNPs, bufferSize, buffer, sections, snpSection, fileLocation, snpLocation, currentSection, dosage, p0, p1, p2, values)
}

